sdks/python/apache_beam/runners/interactive/interactive_beam.py

```diff
@@ -0,0 +1,199 @@
+    options (~apache_beam.options.pipeline_options.PipelineOptions):
+      A configured
+      :class:`~apache_beam.options.pipeline_options.PipelineOptions` object
+      containing arguments that should be used for running the Beam job.
+    argv (List[str]): a list of arguments (such as :data:`sys.argv`)
+      to be used for building a
+      :class:`~apache_beam.options.pipeline_options.PipelineOptions` object.
+      This will only be used if argument **options** is :data:`None`.
+
+  Raises:
+    ~exceptions.ValueError: if either the runner or options argument is not
+      of the expected type.
+  """
+  if not runner:
+    runner = interactive_runner.InteractiveRunner()
+  return beam.Pipeline(runner, options, argv)
+
+
+def visualize(pcoll):
+  """Visualizes a PCollection."""
+  # TODO(ningk)
+  pass
+
+
+def run_pipeline(pipeline, runner=None, options=None):
+  """Runs an iBeam pipeline with selected runner and options.
+
+  When you are done with prototyping in an interactive notebook env, use this
+  method to directly run the pipeline as a normal pipeline on your selected
+  runner.
+  """
+  # TODO(ningk)
+  pass
+
+
+def current_env(cache_manager=None):
+  """Gets current iBeam environment."""
+  global _ibeam_env
+  if not _ibeam_env:
+    _ibeam_env = InteractiveEnv(cache_manager)
+  return _ibeam_env
+
+
+def new_env(cache_manager=None):
+  """Creates a new iBeam environment to replace current one."""
+  global _ibeam_env
+  _ibeam_env = None
+  return current_env(cache_manager)
+
+
+class InteractiveEnv(object):
+  """An interactive environment with cache and pipeline variable metadata.
+
+  iBeam will use the watched variable information to determine if a PCollection
+  is ever assigned to a variable in user pipeline when executing the pipeline
+  and apply magic to automatically cache and use cache for those PCollections
+  if the pipeline is executed within an interactive environment. Users can also
+  visualize and introspect those PCollections in user code since they have
+  handlers to the variables.
+  """
+
+  def __init__(self, cache_manager=None):
+    self._cache_manager = cache_manager
+    # Holds class instances, module object, string of module names.
+    self._watching_set = set()
+    # Holds variables list of (Dict[str, object]).
+    self._watching_dict_list = []
+    # Always watch __main__ module.
+    self.watch('__main__')
+
+  def watch(self, watchable):
+    """Watches a watchable.
+
+    A watchable can be a dictionary of variable metadata such as locals(), a str
+    name of a module, a module object or an instance of a class. The variable
+    can come from any scope even local. Duplicated variable naming doesn't
+    matter since they are different instances. Duplicated variables are also
+    allowed when watching.
+    """
+    if isinstance(watchable, dict):
+      self._watching_dict_list.append(watchable.items())
+    else:
+      self._watching_set.add(watchable)
+
+  def watching(self):
```

Comment:
Could you add pydocs to all public methods?

Reply:
Will do.